### DISCUSSION TITLE
Drizzle plugin - use drizzle objects as type nested in a regular query response.

### DISCUSSION THREAD

**Discussion by olup** [hayes/pothos#1341] [Q&A] [answered]
Basically, I am wondering how reusable are the types generated by drizzle objects. 

I am trying to return an array of regular objects, one key of which would be a drizzle object. I use a regular field (not the drizzle field) on the query and it seems to work ok as long as I only have a single element in my array. 

More than one throws an error from drizzle saying the model can't be found. Am I stretching the plugin too far ?

From what I gather drizzle objects are queried from the db when returning them even if the root payload contains the required fields, which must be because of nested relations' abilities, it's very nice how everything is a single query. But somehow it cannot be performed multiple times, or as a common query, somehow losing the reference to the model ? I hope it makes sense.

**Comment by hayes** [comment]
They should generally be re-usable, and shouldn't result in errors.   The issues you will run into would be more likely related to performance/inefficiency.  

You are likely hitting some kind of a bug.  Can you share more details or a reproduction of the issue?

**Reply by olup** [reply]
Indeed, let me present my situation in more details. I'll just paste some code here but I can try for a small repo - but that would take a few days.

I have a POC for some kind of a RAG API. The resolver lets you do KNN search with pgvector.

I have a schema with a chunk object like such:

```
export const chunk = pgTable("chunk", {
  uuid: uuid().primaryKey().defaultRandom(),
  chunkContent: text().notNull(),
  embedding: vector("embedding", { dimensions: 1536 }),
  resourceUuid: uuid()
    .notNull()
    .references(() => resource.uuid),
});

export const chunkRelations = relations(chunk, ({ one }) => ({
  resource: one(resource, {
    fields: [chunk.resourceUuid],
    references: [resource.uuid],
  }),
}));
```

Then I have this as a pothos drizzle object:
```
const ChunkGqlType = builder.drizzleObject("chunk", {
  fields: (t) => ({
    uuid: t.exposeID("uuid"),
    chunkContent: t.exposeString("chunkContent"),
    resourceUuid: t.exposeID("resourceUuid"),

    resource: t.relation("resource"),
  }),
});
```

Then I have a basic mutation field:
```
    vectorSearch: t.field({
      type: VectorSearchResponse,
      args: {
        input: t.arg({
          type: builder.inputType("vectorSearchInput", {
            fields: (t) => ({
              query: t.string(),
              limit: t.int(),
            }),
          }),
        }),
      },
      resolve: async (root, args, ctx) => {
        const embedding = await getEmbeddings(args.input?.query || "");
        const similarity = sql<number>`1 - (${cosineDistance(
          schema.chunk.embedding,
          embedding
        )})`;
        const similarChunks = await db
          .select({
            uuid: schema.chunk.uuid,
            chunkContent: schema.chunk.chunkContent,
            embedding: schema.chunk.embedding,
            resourceUuid: schema.chunk.resourceUuid,
            similarity,
          })
          .from(schema.chunk)
          .orderBy((t) => desc(t.similarity))
          .limit(args.input?.limit || 4);

        return {
          sources: similarChunks.map((chunk) => ({
            chunk: chunk,
            similarity: chunk.similarity,
          })),
        };
      },
    }),
```

And the VectorSearchResponse object is like this:
```
export const vectorSearchResponse = builder.simpleObject(
  "vectorSearchResponse",
  {
    fields: (t) => ({
      sources: t.field({
        type: [
          builder.simpleObject("source", {
            fields: (t) => ({
              chunk: t.field({
                type: ChunkGqlType,
              }),
              similarity: t.float(),
            }),
          }),
        ],
      }),
    }),
  }
);
```

When I just have one chunk in the response, all goes well, but a soon as I have more than one I am met with this error:

```
ERR Error: Model chunk(fd0b0cd0-1dbb-406f-938f-c92f8ff37eb9) not found
    at <anonymous> (/Users/achanger/dev/experiments/drizzle/node_modules/.pnpm/@pothos+plugin-drizzle@0.5.1_@pothos+core@4.3.0_graphql@16.9.0/node_modules/@pothos/plugin-drizzle/src/model-loader.ts:198:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  path: [ 'vectorSearch', 'sources', 1, 'chunk', 'chunkContent' ],
  locations: [ { line: 6, column: 9 } ],
  extensions: [Object: null prototype] {}
}
```

I have one such error per field and per item returned on top of the first, working one.

**Comment by olup** [comment, marked as answer]
ok, @hayes, found the bug. I think this line: https://github.com/hayes/pothos/blob/8794a128fd40279a0ac9587cc4e64e05795b5264/packages/plugin-drizzle/src/model-loader.ts#L181

should read : `[...entry.models.keys()].map((model) => this.sqlForModel(model))`

**Comment by olup** [comment]
And a PR https://github.com/hayes/pothos/pull/1343
